include/sunmatrix/sunmatrix_ginkgoblock.hpp

```diff
@@ -0,0 +1,399 @@
+ * ----------------------------------------------------------------*/
+
+#ifndef _SUNMATRIX_GINKGOBLOCK_HPP
+#define _SUNMATRIX_GINKGOBLOCK_HPP
+
+#include <cstring>
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/ginkgo.hpp>
+#include <memory>
+#include <sundials/sundials_matrix.hpp>
+#include <sunmatrix/sunmatrix_ginkgo.hpp>
+
+#if (GKO_VERSION_MAJOR < 1) || (GKO_VERSION_MAJOR == 1 && GKO_VERSION_MINOR < 8)
+#error "Ginkgo 1.8.0 or later is required."
+#endif
+
+namespace sundials {
+namespace ginkgo {
+
+using GkoBatchDenseMat = gko::batch::matrix::Dense<sunrealtype>;
+using GkoBatchCsrMat   = gko::batch::matrix::Csr<sunrealtype, sunindextype>;
+using GkoBatchEllMat   = gko::batch::matrix::Ell<sunrealtype, sunindextype>;
+using GkoBatchVecType  = gko::batch::MultiVector<sunrealtype>;
+
+// Forward declare BlockMatrix class
+template<class GkoBatchMatType>
+class BlockMatrix;
+
+namespace impl {
+
+//
+// Prototypes for non-class methods that operate on Matrix
+//
+
+inline std::unique_ptr<GkoBatchVecType> WrapBatchVector(
+  std::shared_ptr<const gko::Executor> gko_exec, gko::size_type num_blocks,
+  N_Vector x);
+
+inline std::unique_ptr<const GkoBatchVecType> WrapConstBatchVector(
+  std::shared_ptr<const gko::Executor> gko_exec, gko::size_type num_blocks,
+  N_Vector x);
+
+template<class GkoBatchMatType>
+void Matvec(BlockMatrix<GkoBatchMatType>& A, GkoBatchVecType* x,
+            GkoBatchVecType* y);
+
+template<class GkoBatchMatType>
+void Matvec(BlockMatrix<GkoBatchMatType>& A, N_Vector x, N_Vector y);
+
+void ScaleAdd(const sunrealtype c, BlockMatrix<GkoBatchDenseMat>& A,
+              BlockMatrix<GkoBatchDenseMat>& B);
+
+void ScaleAdd(const sunrealtype c, BlockMatrix<GkoBatchCsrMat>& A,
+              BlockMatrix<GkoBatchCsrMat>& B);
+
+void ScaleAdd(const sunrealtype c, BlockMatrix<GkoBatchEllMat>& A,
+              BlockMatrix<GkoBatchEllMat>& B);
+
+template<class GkoBatchMatType>
+void ScaleAddI(const sunrealtype c, BlockMatrix<GkoBatchMatType>& A);
+
+template<class GkoBatchMatType>
+void Copy(BlockMatrix<GkoBatchMatType>& A, BlockMatrix<GkoBatchMatType>& B);
+
+//
+// Methods that operate on SUNMatrix
+//
+template<class GkoBatchMatType>
+SUNMatrix_ID SUNMatGetID_GinkgoBlock(SUNMatrix A)
+{
+  return SUNMATRIX_GINKGOBLOCK;
+}
+
+template<class GkoBatchMatType>
+SUNMatrix SUNMatClone_GinkgoBlock(SUNMatrix A)
+{
+  auto Amat{static_cast<BlockMatrix<GkoBatchMatType>*>(A->content)};
+  auto new_mat{new BlockMatrix<GkoBatchMatType>(*Amat)};
+  return new_mat->Convert();
+}
+
+template<class GkoBatchMatType>
+void SUNMatDestroy_GinkgoBlock(SUNMatrix A)
+{
+  auto Amat{static_cast<BlockMatrix<GkoBatchMatType>*>(A->content)};
+  delete Amat; // NOLINT
+  return;
+}
+
+template<class GkoBatchMatType>
+int SUNMatCopy_GinkgoBlock(SUNMatrix A, SUNMatrix B)
+{
+  auto Amat{static_cast<BlockMatrix<GkoBatchMatType>*>(A->content)};
+  Copy(*Amat, *static_cast<ginkgo::BlockMatrix<GkoBatchMatType>*>(B->content));
```

Comment:
I think calling the `Copy` function, which has to be forward declared and then implemented below, makes this more complicated than just putting the one line of code in the body of `Copy` here. Same holds for functions below. Perhaps, it's this way because the template deduction doesn't work otherwise, in which case, ignore this.

Reply:
It's a three line function, so I dont see how it is particularly complicated as-is.